### PR TITLE
WIP: CMake build refactor

### PR DESCRIPTION
**Pull request by mauneyc-LANL**
NOTE: More work needed on corner cases, please test!

Rework of the CMake build toolchain for `singularity-eos`. Some changes from `singularity-opac` carried over.

Highlights

- Simplified "all end-environment" or "all submodule" selection; controlled with `SINGULARITY_USE_SUBMODULES`. If on, no environment checks for dependencies are carried out and all submodules are configured. If off, looks for required/requested dependencies and fails if not found. This felt most efficient, but if needed different use-cases can be appended.
- Updated submodule inclusion; the (likely still underdeveloped) goal here is to modernize how submodules are included. For submodules dependencies _where it is possible_, after calling `add_subdirectory`, the CMake `find_package` call should behave as tho the package is present in the environment, e.g. like it was a loaded via a module or spack. This is part motivated the above, so that once it is determined if the project is built using it's own submods, or system packages, the rest of the code can forget about it.
- Modernized cmake `install` options, e.g. used as `find_package(singularity-eos)` (Still needs work).

Todo:

- Kokkos/EOSPAC test cases. I'll be looking for cracks but there's a lot of egde cases I'll prolly miss.
- Clean up flags and options, there's some redundancy we can polish out.

@Yurlungur @dholladay00 @brryan

### CODE REVIEW

File: cmake/ProcessConfiguration.cmake [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        # TODO: this shouldn't be necessary, tho
+        # i'm leaving it in for now
+        if(NOT TARGET MPI::MPI_CXX)
+            find_package(MPI COMPONENTS CXX QUIET)
+            if(NOT MPI_FOUND)
+                message(FATAL_ERROR "The HDF5 package requires MPI")
+            endif()
+        endif()
+    endif()
+endif()
```
Comment (Yurlungur) [inline cmake/ProcessConfiguration.cmake:88]: newline

File: cmake/SetupCompilers.cmake [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+set(CMAKE_CXX_STANDARD 14)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_EXTENSIONS OFF)
+
+enable_language(CXX)
+include(CMakeDetermineCXXCompiler)
+if(SINGULARITY_USE_FORTRAN)
+  enable_language(Fortran)
+  include(CMakeDetermineFortranCompiler)
+endif()
```
Comment (Yurlungur) [inline cmake/SetupCompilers.cmake:12]: newline

File: cmake/SetupInstall.cmake [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+)
+
+install(FILES "${PROJECT_BINARY_DIR}/${PROJECT_NAME}/Config.cmake"
+     "${PROJECT_BINARY_DIR}/${PROJECT_NAME}ConfigVersion.cmake"
+ DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/${PROJECT_NAME}/cmake
+)
+
+install(DIRECTORY ${PROJECT_SOURCE_DIR}/${PROJECT_NAME}/
+ DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME}
+)
```
Comment (Yurlungur) [inline cmake/SetupInstall.cmake:40]: newline

File: cmake/SubmoduleConfig.cmake [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    set(KokkosKernels_INST_LAYOUTRIGHT OFF CACHE BOOL "" FORCE)
+    set(KokkosKernels_INST_MEMSPACE_HOSTSPACE OFF CACHE BOOL "" FORCE)
+    set(KokkosKernels_INST_OFFSET_INT OFF CACHE BOOL "" FORCE)
+    set(KokkosKernels_INST_OFFSET_SIZE_T OFF CACHE BOOL "" FORCE)
+    set(KokkosKernels_INST_ORDINAL_INT OFF CACHE BOOL "" FORCE)
+    set(KokkosKernels_INST_ORDINAL_INT OFF CACHE BOOL "" FORCE)
+    add_subdirectory(${PROJECT_SOURCE_DIR}/utils/kokkos-kernels)
+else()
+    add_subdirectory(${PROJECT_SOURCE_DIR}/utils/eigen)
+endif()
```
Comment (Yurlungur) [inline cmake/SubmoduleConfig.cmake:41]: newline

**Comment by mauneyc-LANL** [comment]
Comment on d031183

I may have gone overboard in redoing the directory structure, I had a lot of ideas struggling to find resolution. This is a draft of what I perceive as a simpler setup, but it's admittedly a bit muddied at the moment. If there's concern, it's not much work to revert.

On the plus side, with `-DSINGULARITY_USE_KOKKOS=ON -DSINGULARITY_USE_CUDA=ON -DSINGULARITY_USE_KOKKOSKERNELS=ON -DSINGULARITY_BUILD_TESTS=ON`, it builds and passes on darwin p9s. I've not yet rewrote attaching headers from subprojects, which is next on my todos

**Comment by mauneyc-LANL** [comment]
closing as I have redrafted most of this, see #30
